refactor: binary transfer for setObject(int, Object, int)

mkarg opened the pull request:
refactor: binary transfer for setObject(int, Object, int)

Support binary transfer for setObject(int, Object, int) and setObject(int, Object, int, int) and get rid of PGType class hierarchy.

Closes #328

davecramer commented:
Does this work with protocol version 2 ?

davecramer commented on org/postgresql/jdbc2/AbstractJdbc2Statement.java:
```diff
@@ -1866,15 +1847,191 @@ else if (in instanceof InputStream)
+        throw cannotCastException(in.getClass().getName(), "int");
+    }
+
+    private static short castToShort(final Object in) throws SQLException {
```
Performance wise is it better to have different castToShort, castToInt functions with different signatures instead of using instanceof ?

mkarg replied:
@davecramer I do not see what you actually mean. Can you make an example please? As the passed-in value originally is Object I do not see how you want to improve here?

davecramer replied:
I would think the jvm would know the type of the Object. If you provided

castToShort(Boolean in) would the jvm not know to choose that function ?



Dave Cramer

On 5 July 2015 at 17:39, Markus KARG <notifications@github.com> wrote:

> In org/postgresql/jdbc2/AbstractJdbc2Statement.java
> <https://github.com/pgjdbc/pgjdbc/pull/351#discussion_r33899205>:
>
> > +                return ((Number) in).intValue();
> > +            if (in instanceof java.util.Date)
> > +                return (int) ((java.util.Date) in).getTime();
> > +            if (in instanceof Boolean)
> > +                return (Boolean) in ? 1 : 0;
> > +            if (in instanceof Clob)
> > +                return Integer.parseInt(asString((Clob) in));
> > +            if (in instanceof Character)
> > +                return Integer.parseInt(in.toString());
> > +        } catch (final Exception e) {
> > +            throw cannotCastException(in.getClass().getName(), "int", e);
> > +        }
> > +        throw cannotCastException(in.getClass().getName(), "int");
> > +    }
> > +
> > +    private static short castToShort(final Object in) throws SQLException {
>
> @davecramer <https://github.com/davecramer> I do not see what you
> actually mean. Can you make an example please? As the passed-in value
> originally is Object I do not see how you want to improve here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pgjdbc/pgjdbc/pull/351/files#r33899205>.
>

vlsi replied:
@davecramer , You are wrong here.
Java has no multiple dispatch.
In other words, since _compile-time_ variable type is `Object`, runtime would not be able to pick the "correct" overload.
So, either `instanceof` or some hashmap is required.

mkarg replied:
@davecramer To extend Vladimir's answer: Java is statically typed. Hence it is the compiler who decides at compile time which method to invoke, not the JVM at runtime. So your proposal is not possible in Java, unfortunately.

davecramer replied:
Too much groovy coding for me ;)

mkarg replied:
So you're ok with the PR as-is?

mkarg commented:
@davecramer In essence, I just moved around existing code and added additional mappings not supported before (hence it is actually a refactoring mostly). So what actually do you think will _not_ work with protocoll v2?

vlsi commented on org/postgresql/jdbc2/AbstractJdbc2Statement.java:
```diff
@@ -1866,15 +1847,191 @@ else if (in instanceof InputStream)
+            if (in instanceof Number)
+                return ((Number) in).intValue();
+            if (in instanceof java.util.Date)
+                return (int) ((java.util.Date) in).getTime();
```
Is this really expected behavor? Can we dodge "automatic Date to int conversion" and make it fail instead?

davecramer replied:
Looking at the spec, no we don't have to do this, but as this was the current behaviour changing it may break existing code.

mkarg replied:
@davecramer It's up to you. Just tell me if I shall drop Date to int conversion.

davecramer replied:
Leave it in there. It's unlikely to get called

vlsi commented:
@mkarg , 

Regarding the style:

```
+            if (in instanceof String)
+                return Long.parseLong((String) in);
+            if (in instanceof Number)
+                return ((Number) in).longValue();
+            if (in instanceof java.util.Date)
+                return ((java.util.Date) in).getTime();
+            if (in instanceof Boolean)
+                return (Boolean) in ? 1L : 0L;
+            if (in instanceof Clob)
+                return Long.parseLong(asString((Clob) in));
+            if (in instanceof Character)
+                return Long.parseLong(in.toString());
```

Do you think it would be better to use this pattern?
In other words: it optimizes the cases when we can do something sensible, while otherwise it resorts to `Long.valueOf(castToString(in));`, so it is not repeated again and again.
```
+            if (in instanceof Number)
+                return ((Number) in).longValue();
+            if (in instanceof java.util.Date)
+                return ((java.util.Date) in).getTime();
+            if (in instanceof Boolean)
+                return (Boolean) in ? 1L : 0L;
+            return Long.valueOf(castToString(in));
```

mkarg commented:
@vlsi Unfortunately this breaks the backwards behaviour: The original code threw an exception for unknown types, which is not possible with your optimization. Actually I had the code that way in the first place, and needed to "un-optimize" to not break existing applications (which seems to be a top priority for Dave).

vlsi commented:
Ok, :+1: from me.
